internal/provider/resource_gitlab_personal_access_tokens.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	if !currentUserAdmin {
+		return diag.Errorf("[ERROR] Current user needs to be admin when creating a personal access token: %v", err)
+	}
+
+	options := &gitlab.CreatePersonalAccessTokenOptions{
+		Name:   gitlab.String(d.Get("name").(string)),
+		Scopes: stringSetToStringSlice(d.Get("scopes").(*schema.Set)),
+	}
+
+	log.Printf("[DEBUG] create gitlab PersonalAccessToken %s (scopes: %s) for user ID %d", *options.Name, options.Scopes, userId)
+
+	if v, ok := d.GetOk("expires_at"); ok {
+		parsedExpiresAt, err := time.Parse("2006-01-02", v.(string))
+		if err != nil {
+			return diag.Errorf("Invalid expires_at date: %v", err)
+		}
+
+		parsedExpiresAtISOTime := gitlab.ISOTime(parsedExpiresAt)
+		options.ExpiresAt = &parsedExpiresAtISOTime
+	}
+
+	personalAccessToken, _, err := client.Users.CreatePersonalAccessToken(userId, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tokenId := strconv.Itoa(personalAccessToken.ID)
+	userIdString := strconv.Itoa(userId)
+
+	d.SetId(buildTwoPartID(&userIdString, &tokenId))
+	// NOTE: the token can only be read once after creating it
+	d.Set("token", personalAccessToken.Token)
+
+	return resourceGitlabPersonalAccessTokenRead(ctx, d, meta)
+}
+
+func resourceGitlabPersonalAccessTokenRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	userId, tokenId, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.Errorf("Error parsing ID: %s", d.Id())
+	}
+
+	userIdInt, err := strconv.Atoi(userId)
+	if err != nil {
+		return diag.Errorf("Error converting user ID to string: %v", userId)
+	}
+
+	client := meta.(*gitlab.Client)
+
+	personalAccessTokenId, err := strconv.Atoi(tokenId)
+	if err != nil {
+		return diag.Errorf("%s cannot be converted to int", tokenId)
+	}
+
+	log.Printf("[DEBUG] read gitlab PersonalAccessToken %d, user ID %s", personalAccessTokenId, userId)
+
+	personalAccessToken, err := resourceGitlabPersonalAccessTokenFind(ctx, client, userIdInt, personalAccessTokenId)
+	if errors.Is(err, errResourceGitlabPersonalAccessTokenNotFound) {
+		log.Printf("[DEBUG] failed to read gitlab PersonalAccessToken %d, user ID %s", personalAccessTokenId, userId)
+		d.SetId("")
+	}
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.Set("user_id", userIdInt)
+	d.Set("name", personalAccessToken.Name)
+	if personalAccessToken.ExpiresAt != nil {
+		d.Set("expires_at", personalAccessToken.ExpiresAt.String())
+	}
+	d.Set("active", personalAccessToken.Active)
+	d.Set("created_at", personalAccessToken.CreatedAt.Format(time.RFC3339))
+	d.Set("revoked", personalAccessToken.Revoked)
+
+	err = d.Set("scopes", personalAccessToken.Scopes)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceGitlabPersonalAccessTokenDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	_, tokenId, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.Errorf("Error parsing ID: %s", d.Id())
```

Comment:
Let's also give the user hints about the error at hand here:

```suggestion
		return diag.Errorf("Error parsing ID: %s: %v", d.Id(), err)
```

Reply:
This was changed in a previous review comment if you want a change to the abstracted function to do something like this please let me know

Reply:
the way you've changed it to is okay for me.